gemstone/src/swapper/stargate/provider.rs

```diff
@@ -0,0 +1,307 @@
+use std::str::FromStr;
+use std::sync::Arc;
+
+use alloy_core::sol_types::SolCall;
+use alloy_primitives::{hex, Address, Bytes, FixedBytes, U160, U256};
+use async_trait::async_trait;
+use gem_evm::{
+    jsonrpc::{BlockParameter, EthereumRpc, TransactionObject},
+    stargate::contract::{IStargate, MessagingFee, SendParam},
+};
+use primitives::{AssetId, Chain};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    network::{jsonrpc_call, AlienProvider, JsonRpcResult},
+    swapper::{
+        approval::check_approval_erc20, slippage::apply_slippage_in_bp, thorchain::ThorChain, ApprovalType, FetchQuoteData, GemSwapProvider, SwapChainAsset,
+        SwapProvider, SwapProviderData, SwapQuote, SwapQuoteData, SwapQuoteRequest, SwapRoute, SwapperError,
+    },
+};
+
+use super::{
+    endpoint::{StargateEndpoint, STARGATE_ROUTES},
+    layer_zero::scan::LayerZeroScanApi,
+};
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+struct StargateRouteData {
+    send_param: SendParam,
+    fee: MessagingFee,
+    refund_address: String,
+}
+
+#[derive(Debug, Default)]
+pub struct Stargate {
+    pub enpoints: Vec<StargateEndpoint>,
```

Comment:
there are a lot of iter() loops in the implementation, it seems a map is easier to look up by chain